Centralize policy incremental update logic

squeed opened the pull request:
This PR fixes a race condition first noticed by @gandro (it should be said, this race condition does not actually happen in production). This race condition blocks FQDN identity pre-allocation.

We currently have two codepaths used for pushing newly-allocated identities in to the policy engine: the allocator (via the parameter `notifyOwner = true`), and the IPCache, which sets `notifyOwner = false` and handles policy updates manually. We should combine these paths -- both for correctness' sake as well as fixing the latent race condition.

The race condition was that a new identity may still be in-flight while the allocator reports it as "present", thus the ipcache may incorrectly assume it is already fully propagated.

So, this moves the identity update logic out of the allocator in to it's own cell, then updates the ipcache to use it. In doing so, it unifies incremental updates and prevents interleaved updates.

The ipcache relies heavily on policy updates being completely propagated before proceeding with deletions, so the only serious change was to add a done channel to ensure that all in-flight incremental updates have completed. This solves the race condition noticed previously.

(Note to reviewers: review by commit. One commit is just a cut-and-paste with no code changes. The others are small with new functionality.)

squeed committed:
policy/identity updater: handle identity mutation

The SelectorCache cannot guarantee transactional updates when identities
are mutated, rather than being merely created or deleted. In this case,
the caller is responsible for triggering a full policy recalculation.

Add this logic to the identity updater to match the ipcache.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy: move selectorcache updater to explicit cell

The so-called "IdentityAllocatorOwner" is responsible for pushing
newly-allocated identities in to the SelectorCache and distributing
incremental updates. This logic is duplicated between the IPCache and
the identity cache, but it shouldn't be.

This commit extracts the updater to an independent exported cell. The code
is otherwise unchanged.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy/identity updater: return propagation complete channel

This allows for synchronous addition of identities. There are certain
circumstances where we need to wait for full, end-to-end propagation of
incremental identity updates to BPF policy maps. This method exposes
that.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
ipcache: refactor identity allocation

Previously, the ipcache implemented its own duplicated logic for pushing
incremental updates to the policy maps. Switch to the implementation
shared with the identity allocator.

This prevents possible race conditions where the allocator and ipcache
are distributing the same identities simultaneously.

As part of this, switch to AllocateLocalIdentity() and
ReleaseLocalIdentities(). This means that our allocations will be
serialized with any other users of the allocator.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

fristonio reviewed (approved):
🚢

gandro commented on pkg/policy/cell/identity_updater.go:
```diff
@@ -0,0 +1,223 @@
+	UpdateIdentities(added, deleted identity.IdentityMap)
+}
+
+// identityAllocatorOwner is used to break the circular dependency between
```
nit: `identityAllocatorOwner` still refers to the old name

gandro replied:
re-nit: Not resolved (but this is not a blocker)

gandro commented on pkg/policy/cell/identity_updater.go:
```diff
@@ -0,0 +1,223 @@
+	// This synchronously updates the SelectorCache and queues an incremental
+	// update to any selectors. The waitgroup is closed when all endpoints
+	// have been notified.
+	i.policy.GetSelectorCache().UpdateIdentities(added, deleted, wg)
```
This undoes the `mutated` check introduced in the previous commit. This is not intentional, right?

Edit: Ah, I see the next commit re-introduces it. The commit message here is a bit misleading (`The code
is otherwise unchanged.`), but ultimately I think it's fine.

squeed replied:
Oh whoops, that's awkward. I'll clean up the commit history.

gandro replied:
(note that this is also still the same)

gandro commented on pkg/ipcache/metadata.go:
```diff
@@ -337,13 +336,13 @@ func (m *metadata) findAffectedChildPrefixes(parent cmtypes.PrefixCluster) (chil
 //
 // Do not call this directly; rather, use TriggerLabelInjection()
-func (ipc *IPCache) doInjectLabels(ctx context.Context, modifiedPrefixes []cmtypes.PrefixCluster) (remainingPrefixes []cmtypes.PrefixCluster, mutated bool, err error) {
+func (ipc *IPCache) doInjectLabels(ctx context.Context, modifiedPrefixes []cmtypes.PrefixCluster) (remainingPrefixes []cmtypes.PrefixCluster, err error) {
```
nit: The doc comment here still mentions the removed `mutated` return value. Might be worth documenting that this function now also triggers policy regeneration.

gandro replied:
(also not resolved)

gandro commented on pkg/ipcache/metadata.go:
```diff
@@ -639,48 +641,13 @@ func (ipc *IPCache) doInjectLabels(ctx context.Context, modifiedPrefixes []cmtyp
-		if ipc.UpdatePolicyMaps(ctx, nil, idsToDelete) {
-			mutated = true
-		}
+		ipc.IdentityUpdater.UpdateIdentities(nil, idsToDelete)
```
Why are we no longer waiting for deleted identities to propagate? Is this race-free?

squeed replied:
Very good catch! I don't know what I was thinking. We absolutely have to wait here.

squeed replied:
Actually, looking at the code a bit more, we don't have to wait here, but there is a long-standing race condition in the ipcache where we delete prefixes *after* releasing their identity, causing drops when a prefix reverts to world. I'll fix this in a separate PR, as it will need to be backported.

gandro reviewed (commented):
Awesome cleanup, glad to see this happening! A few minor points of feedback

squeed committed:
cache/allocator: fix interleaved release and allocate

This fixes a potential bug when two callers release and acquire the same
identity in short sequence. First of all, it serializes local allocation
behind a lock, ensuring that subsequent releases and acquires never
happen until after the SelectorCache is updated.

Secondly, it exposes a new method, ReleaseLocalIdentities(), which
allows for efficient batched release of a large number of identities.

Otherwise, the following sequence may result in an inconsistent selector
cache:

1. Allocate(lbls=A, notify=True)
2. Release(A, notify=False): deleted
3. Allocate(lbls=A, notify=True); UpdateIdentities(added: A)
4. UpdateIdentities(deleted: A)

Future commits will remove `notifyOwner` from all Release methods.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

gandro reviewed (commented):
I don't see any changes to the previous version?

gandro reviewed (approved):
Awesome work! These parts of IPCache are always a bit scary, but it looks fine to me from what I can tell.

christarazi commented on pkg/identity/cache/allocator.go:
```diff
@@ -134,10 +135,23 @@ type IdentityAllocator interface {
+	// updates the SelectorCache.
+	//
+	// Returns the list of released (refcount = 0) identities
+	ReleaseLocalIdentities([]identity.NumericIdentity) ([]identity.NumericIdentity, error)
```
Optional nit: it might be cleaner to have `ReleaseLocalIdentities(identity.NumericIdentity...)` and that way you can pass a slice like `ids...` or if you have only a single identity to release, then we can avoid a temporary slice instantiation.

squeed replied:
good point, done.

christarazi reviewed (approved):
Nice work, very clean implementation!
